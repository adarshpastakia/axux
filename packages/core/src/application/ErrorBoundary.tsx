/**
 * AxUX React+TailwindCSS UI Framework
 * @author    : Adarsh Pastakia
 * @version   : 2.0.0
 * @copyright : 2022
 * @license   : MIT
 */

import { Component, type ComponentType, type FC, Fragment, useMemo } from "react";
import { useTranslation } from "react-i18next";
import { useGlobals } from "../context/Global";
import { type ChildrenProp } from "../types";

const ErrorIcon: FC = () => (
  <svg
    version="1.1"
    x="0px"
    y="0px"
    viewBox="0 0 511.999 511.999"
    className="w-12 h-12"
  >
    <circle style={{ fill: "#E24C4B" }} cx="227.556" cy="227.556" r="227.556" />
    <path
      style={{ fill: "#D1403F" }}
      d="M455.111,227.556c0,125.156-102.4,227.556-227.556,227.556c-72.533,0-136.533-32.711-177.778-85.333
	c38.4,31.289,88.178,49.778,142.222,49.778c125.156,0,227.556-102.4,227.556-227.556c0-54.044-18.489-103.822-49.778-142.222
	C422.4,91.022,455.111,155.022,455.111,227.556z"
    />
    <path
      style={{ fill: "#FFFFFF" }}
      d="M331.378,331.378c-8.533,8.533-22.756,8.533-31.289,0l-72.533-72.533l-72.533,72.533
	c-8.533,8.533-22.756,8.533-31.289,0c-8.533-8.533-8.533-22.756,0-31.289l72.533-72.533l-72.533-72.533
	c-8.533-8.533-8.533-22.756,0-31.289c8.533-8.533,22.756-8.533,31.289,0l72.533,72.533l72.533-72.533
	c8.533-8.533,22.756-8.533,31.289,0c8.533,8.533,8.533,22.756,0,31.289l-72.533,72.533l72.533,72.533
	C339.911,308.622,339.911,322.844,331.378,331.378z"
    />
  </svg>
);
const DefaultError: FC<{ error?: string }> = ({ error }) => {
  const { t } = useTranslation("core");
  return (
    <Fragment>
      <h4 className="text-xl font-medium">{t("error.title")}</h4>
      <p>{error}</p>
    </Fragment>
  );
};

const ErrorMessage: FC<KeyValue> = ({ error, errorElement }) => {
  const { errorElement: globalError } = useGlobals();
  const EE = useMemo(() => errorElement ?? globalError ?? DefaultError, []);
  return (
    <div className="p-4">
      <div className="flex items-center">
        <ErrorIcon />
        <div className="px-2 self-start flex-auto">
          <EE error={error} />
        </div>
      </div>
    </div>
  );
};

const ErrorStack: FC<KeyValue> = ({ stack }) => {
  return (
    <div className="p-4 overflow-auto border-t">
      <pre>{stack}</pre>
    </div>
  );
};

/**
 * Error boundary wrapper
 */
export class AxErrorBoundary extends Component<
  {
    errorElement?: ComponentType<{ error?: string }>;
    isMinimal?: boolean;
  } & ChildrenProp,
  { hasError: boolean; error?: string; stack?: AnyObject }
> {
  public static getDerivedStateFromError(error: Error) {
    return { hasError: true, error: error.message, stack: error.stack };
  }

  constructor(props: AnyObject) {
    super(props);
    this.state = { hasError: false };
  }

  public componentDidCatch(error: Error) {
    this.setState({ hasError: true, error: error.message, stack: error.stack });
  }

  render() {
    const env = process.env.NODE_ENV;
    if (this.state.hasError) {
      const { errorElement: E, isMinimal = false } = this.props;
      // You can render any custom fallback UI
      return isMinimal && E != null ? (
        <E error={this.state.error} />
      ) : (
        <div className="ax-error-boundary">
          <div>
            <ErrorMessage error={this.state.error} errorElement={E} />
            {env === "development" && <ErrorStack stack={this.state.stack} />}
          </div>
        </div>
      );
    }

    return this.props.children;
  }
}
