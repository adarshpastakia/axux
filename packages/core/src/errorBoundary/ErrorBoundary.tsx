// @author    : Adarsh Pastakia
// @version   : 0.0.1
// @copyright : 2021
// @license   : MIT

import { Component, ComponentType } from "react";
import { Globals } from "../context/Globals";

const ErrorIcon = () => (
  <svg version="1.1" x="0px" y="0px" viewBox="0 0 511.999 511.999" height={48} width={48}>
    <circle style={{ fill: "#E24C4B" }} cx="227.556" cy="227.556" r="227.556" />
    <path
      style={{ fill: "#D1403F" }}
      d="M455.111,227.556c0,125.156-102.4,227.556-227.556,227.556c-72.533,0-136.533-32.711-177.778-85.333
	c38.4,31.289,88.178,49.778,142.222,49.778c125.156,0,227.556-102.4,227.556-227.556c0-54.044-18.489-103.822-49.778-142.222
	C422.4,91.022,455.111,155.022,455.111,227.556z"
    />
    <path
      style={{ fill: "#FFFFFF" }}
      d="M331.378,331.378c-8.533,8.533-22.756,8.533-31.289,0l-72.533-72.533l-72.533,72.533
	c-8.533,8.533-22.756,8.533-31.289,0c-8.533-8.533-8.533-22.756,0-31.289l72.533-72.533l-72.533-72.533
	c-8.533-8.533-8.533-22.756,0-31.289c8.533-8.533,22.756-8.533,31.289,0l72.533,72.533l72.533-72.533
	c8.533-8.533,22.756-8.533,31.289,0c8.533,8.533,8.533,22.756,0,31.289l-72.533,72.533l72.533,72.533
	C339.911,308.622,339.911,322.844,331.378,331.378z"
    />
  </svg>
);

/**
 * Error boundary wrapper
 * @internal
 */
export class ErrorBoundary extends Component<
  { errorElement?: ComponentType<{ error?: string }> },
  { hasError: boolean; error?: string; stack?: AnyObject }
> {
  public static getDerivedStateFromError(error: Error) {
    return { hasError: true, error: error.message, stack: error.stack };
  }

  constructor(props: AnyObject) {
    super(props);
    this.state = { hasError: false };
  }

  public componentDidCatch(error: Error) {
    this.setState({ hasError: true, error: error.message, stack: error.stack });
  }

  render() {
    const env = process.env.NODE_ENV;
    if (this.state.hasError) {
      const { errorElement: E } = this.props;
      // You can render any custom fallback UI
      return (
        <Globals.Consumer>
          {({ errorElement: DefaultMessage }) => (
            <div className="ax-errorBoundary">
              <div>
                <div style={{ padding: 16 }}>
                  <div className="ax-flex ax-row--middle">
                    <ErrorIcon />
                    <div className="ax-padding--x--md ax-align--start ax-col--fill">
                      {E ? (
                        <E error={this.state.error} />
                      ) : (
                        <DefaultMessage error={this.state.error} />
                      )}
                    </div>
                  </div>
                </div>
                {env === "development" && (
                  <>
                    <hr />
                    <div style={{ overflow: "auto", padding: 16 }}>
                      <pre>{this.state.stack}</pre>
                    </div>
                  </>
                )}
              </div>
            </div>
          )}
        </Globals.Consumer>
      );
    }

    return this.props.children;
  }
}
