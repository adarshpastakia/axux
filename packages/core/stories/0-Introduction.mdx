import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

```tsx
import { AxApplicationProvider,AxViewport } from "@axux/core";
import { I18nextProvider } from "react-i18next";
import { BrowserRouter } from "react-router";

export const Application = () => {
  return (
    {/** i18n for localization **/}
    <I18nextProvider>
      {/** react router **/}
      <BrowserRouter>
        <AxApplicationProvider>
          <AxViewport>
            .
            .
            .
            .
          </AxViewport>
        </AxApplicationProvider>
      </BrowserRouter>
    </I18nextProvider>
  );
};
```
