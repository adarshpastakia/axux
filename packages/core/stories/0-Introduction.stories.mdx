import { Title, Canvas, Story } from "@storybook/addon-docs";

<Meta title="@core/Introduction" />

<Title>Core Components</Title>
