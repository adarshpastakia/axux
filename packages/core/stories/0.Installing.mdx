import { Title, Canvas, Story } from "@storybook/addon-docs";
import { Anchor, Subheading } from "@storybook/blocks";
import { Meta } from "@storybook/blocks";

<Meta title="@core/Installing" />

# Core Components

<Anchor storyId="install">
  <Subheading>Installing</Subheading>
</Anchor>

```bash

#npm
npm install @axux/core

# yarn
yarn add @axux/core

```

<Anchor storyId="using">
  <Subheading>Using AxUX</Subheading>
</Anchor>

<cite type="code">src/index.tsx</cite>

```tsx
import { AxApplicationProvider, AxViewport } from "@axux/core";
import { Fragment, StrictMode } from "react";
import ReactDOM from "react-dom/client";
import { I18nextProvider } from "react-i18next";
import { Outlet, RouterProvider, createMemoryRouter } from "react-router";
import i18next from "./i18n";
import reportWebVitals from "./reportWebVitals";
import "./styles.css";

const root = ReactDOM.createRoot(document.getElementById("root"));

const router = createMemoryRouter([
  {
    path: "/",
    element: (
      <AxViewport>
        <div>Build the ui here</div>
        <Outlet />
      </AxViewport>
    ),
    children: [
      {
        path: "page",
        element: <Page />,
      },
      {
        path: "page",
        element: <Page />,
      },
    ],
  },
]);

const Wrapper = process.env.NODE_ENV === "development" ? Fragment : StrictMode;
root.render(
  <Wrapper>
    <AxApplicationProvider defaultPrimary="denim" defaultAccent="coral">
      <I18nextProvider i18n={i18next}>
        <RouterProvider router={router} />
      </I18nextProvider>
    </AxApplicationProvider>
  </Wrapper>,
);

// @ts-expect-error ignore
if (module.hot) {
  // @ts-expect-error ignore
  module.hot.accept();
}

// If you want to start measuring performance in your app, pass a function
// to log results (for example: reportWebVitals(console.log))
// or send to an analytics endpoint. Learn more: https://bit.ly/CRA-vitals
reportWebVitals(console.debug);
```
