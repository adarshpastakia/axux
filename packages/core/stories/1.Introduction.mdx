import {
  Title,
  Canvas,
  Story,
  Description,
  Heading,
  Subheading,
  Anchor,
} from "@storybook/addon-docs";
import { Meta } from "@storybook/blocks";
import {
  AxAnimation,
  AxApplicationProvider,
  AxAside,
  AxBreadcrumb,
  AxButton,
  AxCallout,
  AxCard,
  AxCollapsable,
  AxContent,
  AxDivider,
  AxErrorBoundary,
  AxFooter,
  AxHeader,
  AxFlexBox,
  AxLoader,
  AxIcon,
  AxMenu,
  AxMeter,
  AxSection,
  AxPage,
  AxProgress,
  AxTag,
  AxText,
  AxTheme,
  AxTitle,
  AxViewport,
  useApplicationContext,
} from "../src";

<Meta title="@core/Introduction" />

# Core Components

<Anchor storyId="app">
  <Subheading>Application</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxApplicationProvider</label>
    <Description of={AxApplicationProvider} />
  </div>
  <div className="comp-card">
    <label>AxViewport</label>
    <Description of={AxViewport} />
  </div>
  <div className="comp-card">
    <label>AxPage</label>
    <Description of={AxPage} />
  </div>
  <div className="comp-card">
    <label>AxErrorBoundary</label>
    <Description of={AxErrorBoundary} />
  </div>
</div>

<Anchor storyId="pageLayouts">
  <Subheading>Page Layout</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxSection</label>
    <Description of={AxSection} />
  </div>
  <div className="comp-card">
    <label>AxContent</label>
    <Description of={AxContent} />
  </div>
  <div className="comp-card">
    <label>AxAside</label>
    <Description of={AxAside} />
  </div>
  <div className="comp-card">
    <label>AxHeader</label>
    <Description of={AxHeader} />
  </div>
  <div className="comp-card">
    <label>AxFooter</label>
    <Description of={AxFooter} />
  </div>
</div>

<Anchor storyId="flexLayouts">
  <Subheading>Flex Layout</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxFlexBox</label>
    <Description of={AxFlexBox} />
  </div>
  <div className="comp-card">
    <label>AxFlexBox.Row</label>
    <Description of={AxFlexBox.Row} />
  </div>
  <div className="comp-card">
    <label>AxFlexBox.Col</label>
    <Description of={AxFlexBox.Col} />
  </div>
</div>

<Anchor storyId="typography">
  <Subheading>Typography</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxText</label>
    <Description of={AxText} />
  </div>
  <div className="comp-card">
    <label>AxText.Abbr</label>
    <Description of={AxText.Abbr} />
  </div>
  <div className="comp-card">
    <label>AxText.Mark</label>
    <Description of={AxText.Mark} />
  </div>
  <div className="comp-card">
    <label>AxText.Copy</label>
    <Description of={AxText.Copy} />
  </div>
  <div className="comp-card">
    <label>AxTitle</label>
    <Description of={AxTitle} />
  </div>
</div>

<Anchor storyId="buttons">
  <Subheading>Buttons</Subheading>
</Anchor>

<div className="card-grid">
  <div className="comp-card">
    <label>AxButton</label>
    <Description of={AxButton} />
  </div>
  <div className="comp-card">
    <label>AxButton.Group</label>
    <Description of={AxButton.Group} />
  </div>
  <div className="comp-card">
    <label>AxButton.Confirm</label>
    <Description of={AxButton.Confirm} />
  </div>
  <div className="comp-card">
    <label>AxButton.Action</label>
    <Description of={AxButton.Action} />
  </div>
  <div className="comp-card">
    <label>AxButton.Dropdown</label>
    <Description of={AxButton.Dropdown} />
  </div>
</div>
