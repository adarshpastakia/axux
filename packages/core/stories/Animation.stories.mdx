import {
  Anchor,
  ArgsTable,
  Canvas,
  Description,
  Heading,
  Meta,
  Story,
  Title,
} from "@storybook/addon-docs";
import { useState, useEffect } from "react";
import { AxAnimation, AxBadge, AxProgress } from "../src";

<Meta title="@core/Animations" />

<Title>Animations</Title>

<Heading storyId="core-animations--1a-check">Animated Circles</Heading>

<Canvas>
  <Story name="1a.Check">
    <AxAnimation.Check className="text-success-500" />
  </Story>
  <Story name="1b.Cross">
    <AxAnimation.Cross className="text-danger-500" />
  </Story>
  <Story name="1c.Info">
    <AxAnimation.Info className="text-primary-500" />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Check} />

<Heading storyId="core-animations--2-a-bars">Animated Indicators</Heading>

<Canvas>
  <Story name="2a.Bars">
    <AxAnimation.Bars className="text-primary-500" />
  </Story>
  <Story name="2b.Spinner">
    <AxAnimation.Spinner className="text-4xl text-accent-500" />
  </Story>
  <Story name="2c.Card">
    <AxAnimation.Card className="w-64" showIcon />
  </Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxAnimation.Bars} />

<Heading storyId="core-animations--3-a-progress-circle">Progress</Heading>

export const CircleTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return <AxProgress.Circle size="md" value={value} />;
};
const CircleStory = CircleTemplate.bind({});
CircleStory.args = {};

export const BarTemplate = (args) => {
  const [value, setValue] = useState(24);
  useEffect(() => {
    const tmr = setInterval(() => {
      setValue(Math.random() * 100);
    }, 5000);
    return () => {
      clearInterval(tmr);
    };
  }, []);
  return (
    <div style={{ width: "40rem" }}>
      <AxProgress.Bar size="md" value={value} animate />
    </div>
  );
};
const BarStory = BarTemplate.bind({});
BarStory.args = {};

<Canvas>
  <Story name="3a.Progress Circle">{CircleStory}</Story>
  <Story name="3b.Progress Bar">{BarStory}</Story>
</Canvas>

<ArgsTable sort="requiredFirst" of={AxProgress.Bar} />
