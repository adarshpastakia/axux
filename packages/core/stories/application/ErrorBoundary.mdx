import {
  ArgsTable,
  Canvas,
  Description,
  Primary,
  Story,
  Title,
} from "@storybook/addon-docs";
import {
  Anchor,
  ArgTypes,
  Controls,
  Meta,
  Subheading,
} from "@storybook/blocks";
import {
  AxApplicationProvider,
  AxCollapsable,
  AxErrorBoundary,
  AxPage,
  AxViewport,
  useApplicationContext,
} from "../../src";
import * as ErrorBoundaryStories from "./ErrorBoundary.stories";

<Meta of={ErrorBoundaryStories} />

<Anchor storyId="page">
  <div className="invisible h-0 overflow-hidden">
    <Subheading>Error Boundary</Subheading>
  </div>
</Anchor>

# Error Boundary

<Description of={AxErrorBoundary} />

<Primary />

<AxCollapsable>
  <div>Controls</div>
  <Controls of={ErrorBoundaryStories.Example} />
</AxCollapsable>

<hr />

<Anchor storyId="props">
  <Subheading>Props</Subheading>
</Anchor>

<ArgTypes sort="requiredFirst" of={AxErrorBoundary} />

<hr />

<Anchor storyId="sample">
  <Subheading>Example</Subheading>
</Anchor>

```tsx
import { AxPage, AxHeader, AxContent, AxErrorBoundary } from "@axux/core";

export const RoutePage = () => {
  return (
    <AxPage isPaper isLoading={loadingState} title="Route title">
      <AxHeader>Page header controls</AxHeader>
      <AxContent>
        <AxErrorBoundary
          errorElement={({ error }) => (
            <div>
              <p>Application error [{error}]</p>
              <a className="link">Navigate to previous state</a>
            </div>
          )}
        >
          <BadComponent />
        </AxErrorBoundary>
        <p>Page content</p>
      </AxContent>
    </AxPage>
  );
};
```
