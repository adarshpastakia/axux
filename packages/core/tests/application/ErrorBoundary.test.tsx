/**
 * AxUX React UI Framework with Tailwind CSS
 * @author    : Adarsh Pastakia
 * @version   : 4.0.0
 * @copyright : 2024
 * @license   : MIT
 */

import { composeStories } from "@storybook/testing-react";
import { cleanup, render } from "@testing-library/react";
import React from "react";
import * as stories from "../../stories/application/ErrorBoundary.stories";

const { Example } = composeStories(stories);

describe("ErrorBoundary", () => {
  beforeAll(() => {
    render(<Example />);
  });

  it("should render", () => {
    expect(document.querySelector(".ax-error-boundary")).not.toBeNull();
    expect(document.body.innerHTML).toMatchSnapshot();
  });

  afterAll(() => {
    cleanup();
  });
});
