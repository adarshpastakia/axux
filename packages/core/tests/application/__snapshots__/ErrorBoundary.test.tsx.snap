// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorBoundary should render 1`] = `"<div><div class="contents" data-color-scheme="light" data-primary-scheme="denim" data-accent-scheme="jade"><div class="viewport-wrapper"><div data-test-loading="false" class="ax-page  "><div class="ax-section "><div role="none" class="ax-header text-primary justify-center">Title</div><div class="ax-content "><div class="ax-error-boundary"><div><div class="p-4"><div class="flex items-center"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 64 64" class="size-10"><path fill="#242424" d="M64 32c0 17.673-14.327 32-32 32s-32-14.327-32-32 14.327-32 32-32 32 14.327 32 32z"></path><path fill="#ffffff" d="M63.5 32c0 17.397-14.103 31.5-31.5 31.5s-31.5-14.103-31.5-31.5 14.103-31.5 31.5-31.5 31.5 14.103 31.5 31.5z"></path><path fill="#f44336" d="M60 32c0 15.464-12.536 28-28 28s-28-12.536-28-28c0-15.464 12.536-28 28-28s28 12.536 28 28z"></path><path fill="#fafafa" d="M50.667 35.733h-37.333c-1.031 0-1.867-0.835-1.867-1.867v-3.733c0-1.031 0.835-1.867 1.867-1.867h37.333c1.031 0 1.867 0.835 1.867 1.867v3.733c0 1.031-0.835 1.867-1.867 1.867z"></path></svg><div class="px-2 self-start flex-auto"><div><p>Application error [badcall is not defined]</p><a class="link">Navigate to previous state</a></div></div></div></div></div></div></div></div></div></div><div class="ax-overlay__container" data-mode="overlay"></div><div class="ax-overlay__container" data-mode="alert"></div><div class="ax-notification__container" data-mode="toast"></div><div class="ax-notification__container" data-mode="message"></div></div></div>"`;
